gateway-manager/src/manage_confluent_cloud.py

```diff
@@ -0,0 +1,605 @@
+# you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+from dataclasses import dataclass, fields
+from datetime import datetime
+import re
+import sys
+
+import pexpect
+
+import helpers
+from settings import (
+    CC_CLI_PATH,
+    CC_API_USER,
+    CC_API_PASSWORD,
+    CC_CLUSTER_NAME,
+    CC_ENVIRONMENT_NAME
+)
+
+
+LOGGER = helpers.get_logger('ConfluentCloud')
+
+
+# field order MUST match return order for CCloud
+@dataclass
+class Environment:
+    id: str
+    name: str
+
+
+@dataclass
+class Cluster:
+    id: str
+    name: str
+    provider: str
+    region: str
+    durability: str
+    status: str
+
+
+@dataclass
+class ServiceAccount:
+    id: int
+    name: str
+    desc: str
+
+
+@dataclass
+class NewAPIKey:
+    name: str
+    key: str
+
+
+@dataclass
+class APIKey:
+    id: str
+    service_account_id: int
+    desc: str
+
+
+@dataclass
+class ACL:
+    id: int  # ID of ServiceAccount
+    permission: str
+    operation: str
+    resource_class: str
+    resource_id: str
+    acl_type: str
+
+
+#####################################
+#
+#   Access
+#
+#####################################
+
+def login(user, pw):
+    login_prompt = 'Email:*'
+    pw_prompt = 'Password:*'
+    child = pexpect.spawn(f'{CC_CLI_PATH} login --url https://confluent.cloud')
+    child.expect(login_prompt)
+    child.sendline(user)
+    child.expect(pw_prompt)
+    child.sendline(pw)
+    res = child.read().decode("utf-8")
+    if 'Logged in' not in res:
+        raise ConnectionError(f'Login Failed: {_inline(res)}')
+    LOGGER.debug(_inline(res))
+    return True
+
+
+def logout():
+    child = pexpect.spawn(f'{CC_CLI_PATH} logout')
+    res = child.read().decode("utf-8")
+    if 'You are now logged out' not in res:
+        raise LOGGER.error(f'Could not disconnect: {_inline(res)}')
+    LOGGER.debug(_inline(res))
+    return True
+
+
+#####################################
+#
+#   Environments
+#
+#####################################
+
+
+def _handle_env(i):
+    # removes the currently selected indicator (*) from the name
+    i['id'] = i['id'].replace('*', '').strip()
+    return i
+
+
+def list_environments():
+    LOGGER.debug('Fetching environments...')
+    child = pexpect.spawn(f'{CC_CLI_PATH} environment list')
+    res = str(child.read())
+    return _process_resource(Environment, res, _handle_env)
+
+
+def set_environment(name):
+    envs = list_environments()
+    match = [e for e in envs if e.name == name]
+    if not match:
+        raise ValueError(f'No environment named: {name}')
```

Comment:
You love the `ValueError` exception, don't you? :stuck_out_tongue: 
But I still think this is not the proper use for it

Reply:
I thought about you as I was writing this. In this case I'm not sure what a better exception is, but I'm open to a suggestion.

Reply:
The generic one for those cases could be `RuntimeError` because the code is not wrong, the value is not wrong, the problem is that the environment value is missing. You have the `EnvironmentError` https://docs.python.org/3/library/exceptions.html#EnvironmentError if you prefer.